plugin/src/main/java/com/playmonumenta/redissync/RedisLock.java

```diff
@@ -0,0 +1,296 @@
+		)
+		.map(response -> "OK".equals(response))
+		.orElse(false);
+	}
+
+	/**
+	 * Gets the shard holding the inter-shard lock.
+	 * Runs watch before getting the shard.
+	 *
+	 * @return The entry in redis at the key
+	 * corresponding to this lock, or None if it
+	 * does not exist.
+	 */
+	private Optional<String> watchThenGetLockingShard() {
+		RedisAPI.getInstance()
+			.async()
+			.watch(mKeyName);
+		return Optional.ofNullable(
+			RedisAPI.getInstance()
+				.sync()
+				.get(mKeyName)
+		);
+	}
+
+	public class RedisLockException extends RuntimeException {
+		public RedisLockException() {
+			super();
+		}
+
+		public RedisLockException(String message) {
+			super(message);
+		}
+	}
+
+	/**
+	 * Attempts to acquire the lock, and blocks this thread
+	 * until it is able to do so.
+	 *
+	 * Upon return, this thread will own both the intra-shard lock
+	 * and the inter-shard lock.
+	 */
+	public void lock() {
+		mInternalAccessLock.lock();
+		try {
+			while (mLockOwner.isPresent()) {
+				mIsIntraLockedCondition.awaitUninterruptibly();
+			}
+			mLockOwner = Optional.of(Thread.currentThread());
+
+			// Block until inter-shard lock is free
+			RedisAPI.getInstance().asyncPubSub().subscribe("__keyspace@0__:" + mKeyName);
+			while (!tryAcquireRedis()) {
+				mIsInterLockedCondition.awaitUninterruptibly();
+			}
+			RedisAPI.getInstance().asyncPubSub().unsubscribe("__keyspace@0__:" + mKeyName);
+
+			// TODO: try catch RejectedExecutionException; gracefully unwind
+			mRefreshTask = Optional.of(
+				mRefreshScheduler.scheduleAtFixedRate(
+					() -> {
+						// Create a new thread so separate refresh tasks
+						// don't get stuck behind each other.
+						mRefreshWorkers.submit(() -> {
+							// Mind the case: key is deleted before this task is shut down,
+							// this task's last run would refresh a different shard's lock.
+
+							// The lifetime of this refresh task could outlive a critical section
+
+							/**
+							 * Two outcomes:
+							 * 
+							 * Found self:
+							 * WATCH
+							 * self_shard = GET mKeyName
+							 * MULTI
+							 * PEXPIRE
+							 * EXEC
+							 * - You know that if the key is self
+							 * - then changes, it must be non-self;
+							 * - do not refresh nonself.
+							 *
+							 * Found else:
+							 * WATCH
+							 * other_shard = GET mKeyName
+							 * UNWATCH
+							 * - If the other shard changes to you,
+							 * - That means you have acquired the lock
+							 * - A separate refresh task has started.
+							 */
+							Optional<String> lockingShard = watchThenGetLockingShard();
+							if (lockingShard.isEmpty() || !lockingShard.get().equals(ConfigAPI.getShardName())) {
+								RedisAPI.getInstance()
+									.async()
+									.unwatch();
+								return;
+							}
+							RedisAPI.getInstance()
+								.async()
+								.multi();
+							RedisAPI.getInstance()
+								.async()
+								.pexpire(mKeyName, mTimeoutMS);
```

Comment:
Is there any reason to use a multi/exec block for a single command? This seems like something that could be done directly instead.